File: backend/query_engine.py
"""QueryEngine (ADR-0001): the RAG chain on LangChain + Groq.

Flow: detect language + translate to English (Groq) -> similarity search in
the Knowledge Base -> answer grounded in retrieved context, replied in the
user's language. Model: openai/gpt-oss-120b via the plain langchain-openai
client pointed at Groq's OpenAI-compatible base URL (no langchain-groq).
"""

import json
import os
from datetime import datetime
from zoneinfo import ZoneInfo

from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

GROQ_BASE_URL = "https://api.groq.com/openai/v1"
MODEL = "openai/gpt-oss-120b"

TRANSLATE_PROMPT = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            "You are a translator.\n"
            "1. Detect the language of the user's message.\n"
            "2. Translate it to English.\n"
            '3. Reply ONLY with JSON: {{"translation": "...", "detected_language": "..."}}',
        ),
        ("human", "{query}"),
    ]
)

ANSWER_PROMPT = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            "You are a helpful assistant for a college.\n"
            "Answer the user's question based strictly on the provided context.\n"
            "Respond in this language: {language}.\n"
            "Current date and time in Asia/Kolkata (use to resolve 'today', 'tomorrow', 'yesterday', weekday names): {current_date}.\n"
            "If the question is about 'today', use the date above to pick the correct weekday from the context (e.g., Monday-Sunday menu).\n"
            "Use INR (Indian Rupees) for all monetary values.\n"
            "If the info is missing from the context, say so politely in {language}.",
        ),
        ("human", "Context:\n{context}\n\nQuestion: {question}\n\nAnswer:"),
    ]
)


class QueryEngine:
    def __init__(self, knowledge_base):
        api_key = os.getenv("GROQ_API_KEY")
        if not api_key:
            raise ValueError("GROQ_API_KEY not found in environment variables")

        self.knowledge_base = knowledge_base
        self.llm = ChatOpenAI(
            model=MODEL,
            api_key=api_key,
            base_url=GROQ_BASE_URL,
            temperature=0.1,
        )
        self.translate_chain = TRANSLATE_PROMPT | self.llm | StrOutputParser()
        self.answer_chain = ANSWER_PROMPT | self.llm | StrOutputParser()

    def process_query(self, user_query: str, history: list[dict] = None) -> tuple[str, list[str]]:
        """
        Returns (answer, sources) where sources is a list of unique filenames
        that contributed to the answer.
        """
        try:
            # Step 1: detect language and translate to English
            english_query, language = self._translate(user_query)

            # Step 2: retrieve context from the Knowledge Base with sources
            context_chunks, source_filenames = self.knowledge_base.search_with_sources(
                english_query[:1000]
            )
            if not context_chunks:
                return (
                    "I couldn't find any relevant documents to answer your question.",
                    [],
                )

            # Build conversation context from recent history (last 5 pairs)
            history_str = ""
            if history and len(history) > 0:
                # Last 5 in chronological order (oldest first in string)
                recent = history[-5:]
                pairs = []
                for h in recent:
                    role = h.get("role", "user")
                    content = h.get("content", "")
                    label = "User" if role == "user" else "Assistant"
                    pairs.append(f"{label}: {content}")
                history_str = "Past conversation:\n" + "\n".join(pairs) + "\n\n"

            # Step 3: generate a grounded answer in the user's language
            # Inject current date for 'today' resolution (Asia/Kolkata)
            current_date = datetime.now(ZoneInfo("Asia/Kolkata")).strftime(
                "%A, %Y-%m-%d %H:%M %Z"
            )
            full_context = ("\n\n".join(context_chunks))
            if history_str:
                full_context = history_str + full_context
            answer = self.answer_chain.invoke(
                {
                    "context": full_context,
                    "question": user_query,
                    "language": language,
                    "current_date": current_date,
                }
            )
            # Deduplicate sources while preserving order
            unique_sources = list(dict.fromkeys(source_filenames))
            return answer or "Sorry, something went wrong. Please try again later.", unique_sources
        except Exception as exc:
            print(f"CRITICAL ERROR in QueryEngine: {exc}")
            return (
                "Sorry, the assistant can't reach its AI service right now. "
                "Please try again later.",
                [],
            )

    def _translate(self, user_query: str) -> tuple[str, str]:
        """Returns (english_query, detected_language); falls back to the original."""
        try:
            raw = self.translate_chain.invoke({"query": user_query})
            clean = raw.replace("```json", "").replace("```", "").strip()
            data = json.loads(clean)
            return data.get("translation", user_query), data.get(
                "detected_language", "English"
            )
        except Exception as exc:
            print(f"Translation failed, using original query: {exc}")
            return user_query, "English"

File: src/lib/query-engine.ts
import { chat, chatJSON } from "./llm";
import { getCollection } from "./database";
import { TranslationResult } from "@/types";

const SYSTEM_INSTR_TRANS = `You are a translator.
1. Detect the language of the query.
2. Translate it to English.
3. Return JSON: {"translation": "...", "detected_language": "..."}`;

function getSystemInstrGen(targetLang: string): string {
  return `You are a helpful assistant for a college.
Answer the user's question based strictly on the provided context.
Respond in the user's detected language: ${targetLang}.
If the info is missing, say so politely in ${targetLang}.`;
}

export async function processQuery(userQuery: string): Promise<string> {
  try {
    let englishQuery = userQuery;
    let userLang = "English";

    try {
      const translationRaw = await chatJSON(SYSTEM_INSTR_TRANS, userQuery);
      const cleaned = translationRaw
        .replace(/```json/g, "")
        .replace(/```/g, "")
        .trim();
      const result: TranslationResult = JSON.parse(cleaned);

      if (result.translation) {
        englishQuery = result.translation;
        userLang = result.detected_language || "English";
      }
    } catch {
      englishQuery = userQuery;
      userLang = "English";
    }

    console.log(
      `Debug: Original: ${userQuery}, Translated: ${englishQuery}, Lang: ${userLang}`
    );

    const safeQuery = englishQuery.slice(0, 1000);
    const contextText = await retrieveContext(safeQuery);

    if (!contextText) {
      return "I couldn't find any relevant documents to answer your question.";
    }

    const userContent = `Context:\n${contextText}\n\nUser Question: ${userQuery}\n\nAnswer:`;
    const finalAnswer = await chat(
      getSystemInstrGen(userLang),
      userContent,
      0.1
    );

    if (!finalAnswer) {
      return "Sorry, the AI model is currently overloaded. Please try again later.";
    }

    return finalAnswer;
  } catch (err) {
    console.error("CRITICAL ERROR in QueryEngine:", err);
    return "Sorry, something went wrong while processing your query. Please try again later.";
  }
}

async function retrieveContext(query: string): Promise<string> {
  try {
    const collection = getCollection();
    const results = collection.find(
      {},
      {
        sort: { $vectorize: query },
        limit: 5,
        projection: { content: 1 },
      }
    );

    const documents: string[] = [];
    for await (const doc of results) {
      if (doc.content) {
        documents.push(doc.content);
      }
    }

    return documents.join("\n\n");
  } catch (err) {
    console.error("Search error in Astra DB:", err);
    return "";
  }
}
